src/ports/postgres/modules/utilities/pivot.py_in

```diff
@@ -0,0 +1,200 @@
+# specific language governing permissions and limitations
+# under the License.
+
+# Pivoting
+# The goal of the MADlib pivot function is to provide a data summarization tool
+# that can do basic OLAP type operations on data stored in one table and output
+# the summarized data to a second table.  Typical operations are count, average,
+# min, max and standard deviation, however user defined aggregates (UDAs) are
+# also be allowed.
+
+# Please refer to the pivot.sql_in file for the documentation
+
+"""
+@file pivot.py_in
+
+"""
+import plpy
+from utilities import _assert
+from utilities import split_quoted_delimited_str
+from utilities import strip_end_quotes
+from validate_args import table_exists
+from validate_args import columns_exist_in_table
+from validate_args import table_is_empty
+from validate_args import _get_table_schema_names
+from validate_args import get_first_schema
+
+m4_changequote(`<!', `!>')
+
+def pivot(schema_madlib, source_table, out_table,
+                 index, pivot_cols, pivot_values,
+                 aggregate_func, **kwargs):
+    """
+    Helper function that can be used to pivot tables
+    Args:
+        @param source_table     The original data table
+        @param out_table        The output table that contains the dummy
+                                variable columns
+        @param index            The index columns to group by the records by
+        @param pivot_cols       The columns to pivot the table
+        @param pivot_values     The value columns to be summarized in the
+                                pivoted table
+        @param aggregate_func   The aggregate function to be applied to the
+                                values
+
+    """
+    indices = split_quoted_delimited_str(index)
+    cols = split_quoted_delimited_str(pivot_cols)
+    vals = split_quoted_delimited_str(pivot_values)
+    # aggregate_func = "sum"
+    validate_pivot_coding(source_table, out_table, indices, cols, vals)
+    new_col_names =[]
+    sql_list = ["CREATE TABLE " + out_table + " AS (SELECT " + index]
+    # Preperation for multiple index, pivot, etc.
+    for col in cols:
+        for val in vals:
+            col_no_quotes = strip_end_quotes(col.strip())
+            val_no_quotes = strip_end_quotes(val.strip())
+            distinct_values = plpy.execute(
+                "SELECT {col} AS value FROM {source_table} "
+                "GROUP BY {col} ORDER BY {col}".
+                format(col=col, source_table=source_table))
+            distinct_values = [strip_end_quotes(item['value']) for item in distinct_values]
+            null_wrap_case_str = ""
+            null_wrap_end_str = ""
+            if None in distinct_values:
+                null_wrap_case_str = "CASE WHEN \"{col}\" is NULL THEN NULL ELSE "
+                null_wrap_end_str = "END"
+            case_str = ("{agg}({wrap_case} "
+                        "CASE WHEN \"{{col}}\" = '{{value}}' THEN {val} ELSE 0 END "
```

Comment:
You'll have to have a programmable default value for other aggs. `0` works for sum but would fail for product. I guess `NULL` is a feasible candidate for that expression.